pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+    """
+
+    @classmethod
+    @Appender(_interval_shared_docs['from_tuples'] % _shared_docs_kwargs)
+    def from_tuples(cls, data, closed='right', copy=False, dtype=None):
+        if len(data):
+            left, right = [], []
+        else:
+            left = right = data
+
+        for d in data:
+            if isna(d):
+                lhs = rhs = np.nan
+            else:
+                name = cls.__name__
+                try:
+                    # need list of length 2 tuples, e.g. [(0, 1), (1, 2), ...]
+                    lhs, rhs = d
+                except ValueError:
+                    msg = ('{name}.from_tuples requires tuples of '
+                           'length 2, got {tpl}').format(name=name, tpl=d)
+                    raise ValueError(msg)
+                except TypeError:
+                    msg = ('{name}.from_tuples received an invalid '
+                           'item, {tpl}').format(name=name, tpl=d)
+                    raise TypeError(msg)
+                lhs, rhs = d
+            left.append(lhs)
+            right.append(rhs)
+
+        return cls.from_arrays(left, right, closed, copy=False,
+                               dtype=dtype)
+
+    def _validate(self):
+        """Verify that the IntervalIndex is valid.
+
+        Checks that
+
+        * closed is valid
+        * left and right match lengths
+        * left and right have the same missing values
+        * left is always below right
+        """
+        if self.closed not in _VALID_CLOSED:
+            raise ValueError("invalid option for 'closed': {closed}"
+                             .format(closed=self.closed))
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        left_mask = notna(self.left)
+        right_mask = notna(self.right)
+        if not (left_mask == right_mask).all():
+            raise ValueError('missing values must be missing in the same '
+                             'location both left and right sides')
+        if not (self.left[left_mask] <= self.right[left_mask]).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    # ---------
+    # Interface
+    # ---------
+    def __iter__(self):
+        return iter(self.values)
+
+    def __len__(self):
+        return len(self.left)
+
+    def __getitem__(self, value):
+        mask = self.isna()[value]
+        if is_scalar(mask) and mask:
+            return self._fill_value
+
+        left = self.left[value]
+        right = self.right[value]
+
+        # scalar
+        if not isinstance(left, Index):
+            return Interval(left, right, self.closed)
+
+        return self._shallow_copy(left, right)
+
+    def __setitem__(self, key, value):
+        if not (is_interval_dtype(value) or isinstance(value, Interval)):
+            msg = "'value' should be an interval type, got {} instead."
+            raise TypeError(msg.format(type(value)))
```

Comment:
Looks like generally `type(value).__name__` is used in messages instead of just `type(value)`.  Not super concerned with this distinction though, just a minor inconsistency I noticed.